File: platform/fastify_server/config/lifecycle/02-create_shared_context.lifecycle.mjs
/**
 * 02-create_shared_context.mjs
 *
 * Creates and decorates the server with SharedContext before context resolver runs.
 * This allows other plugins/modules to pre-configure the shared context.
 *
 * SharedContext Features:
 * - .get(key, default) - get value, callable defaults are invoked and cached
 * - .set(key, value) - set a value
 * - .register(key, value) - register utilities accessible by child contexts
 * - .createChild() - create REQUEST context that inherits from STARTUP context
 */

let createSharedContext;
let HAS_SHARED_CONTEXT = false;

try {
  const resolver = await import("app_yaml_overwrites");
  createSharedContext = resolver.createSharedContext;
  HAS_SHARED_CONTEXT = !!createSharedContext;
} catch (e) {
  console.warn("[lifecycle:create_shared_context] app-yaml-overwrites not available:", e.message);
}

export async function onStartup(server, config) {
  server.log.info(
    { HAS_SHARED_CONTEXT },
    "[lifecycle:create_shared_context] Initializing SharedContext..."
  );

  try {
    if (!HAS_SHARED_CONTEXT) {
      server.log.warn(
        "[lifecycle:create_shared_context] createSharedContext not available. Skipping SharedContext setup."
      );
      return;
    }

    // Check if sharedContext already exists
    if (server.hasDecorator("sharedContext") && server.sharedContext) {
      server.log.info(
        "[lifecycle:create_shared_context] SharedContext already exists on server, skipping creation"
      );
      return;
    }

    // Create and decorate shared context
    server.log.info("[lifecycle:create_shared_context] Creating SharedContext instance");
    const sharedContext = createSharedContext();
    server.decorate("sharedContext", sharedContext);
    server.log.info("[lifecycle:create_shared_context] Created and decorated server with SharedContext");

    // ==========================================================================
    // EXAMPLES: Register utilities at STARTUP
    // ==========================================================================

    // Example 1: Register a simple value
    // sharedContext.set('app_start_time', Date.now());

    // Example 2: Register a utility class instance
    // class TokenGenerator {
    //     generate() {
    //         return `token_${Date.now()}_${Math.random().toString(36).slice(2)}`;
    //     }
    // }
    // sharedContext.register('token_generator', new TokenGenerator());

    // Example 3: Register a lazy-computed value (computed on first access)
    // sharedContext.register('db_connection', () => createDbConnection(), { lazy: true });

    // Example 4: Access values later with callable default (auto-cached)
    // const timestamp = sharedContext.get('timestamp', () => Date.now());
    // const sameTimestamp = sharedContext.get('timestamp', () => Date.now()); // Returns cached value

    server.log.info("[lifecycle:create_shared_context] SharedContext initialized successfully");
  } catch (err) {
    server.log.error({ err, hookName: "02-create_shared_context" }, "[lifecycle:create_shared_context] SharedContext initialization failed");
    throw err;
  }
}

File: platform/fastapi_server/config/lifecycle/02_create_shared_context.lifecycle.py
"""
02_create_shared_context.py

Creates and stores SharedContext on app.state before context resolver runs.
This allows other plugins/modules to pre-configure the shared context.

SharedContext Features:
- .get(key, default) - get value, callable defaults are invoked and cached
- .set(key, value) - set a value
- .register(key, value) - register utilities accessible by child contexts
- .create_child() - create REQUEST context that inherits from STARTUP context

Per-request child contexts:
    Use a FastAPI dependency to get a child context per request:

    from fastapi import Depends, Request

    def get_request_context(request: Request):
        shared = request.app.state.sharedContext
        return shared.create_child() if shared else None

    @app.get("/example")
    async def example(ctx = Depends(get_request_context)):
        value = ctx.get('some_key', lambda: 'default')
        ...
"""

import logging
from fastapi import FastAPI

logger = logging.getLogger("lifecycle:create_shared_context")

try:
    from app_yaml_overwrites import create_shared_context
    HAS_SHARED_CONTEXT = True
except ImportError:
    HAS_SHARED_CONTEXT = False


async def onStartup(app: FastAPI, config: dict):
    """Create and store SharedContext on app.state."""
    logger.info("Starting create_shared_context lifecycle hook...")
    try:
        logger.debug(
            "02_create_shared_context.py onStartup called. HAS_SHARED_CONTEXT=%s",
            HAS_SHARED_CONTEXT,
        )

        if not HAS_SHARED_CONTEXT:
            logger.warning("create_shared_context not available. Skipping SharedContext setup.")
            return

        # Check if sharedContext already exists
        existing = getattr(app.state, 'sharedContext', None)
        if existing is not None:
            logger.info("SharedContext already exists on app.state, skipping creation")
            return

        # Create and store shared context
        shared_context = create_shared_context()
        app.state.sharedContext = shared_context
        logger.info("Created and stored SharedContext on app.state")

        # ==========================================================================
        # EXAMPLES: Register utilities at STARTUP
        # ==========================================================================

        # Example 1: Register a simple value
        # import time
        # shared_context.set('app_start_time', time.time())

        # Example 2: Register a utility class instance
        # class TokenGenerator:
        #     def generate(self):
        #         import uuid
        #         return f"token_{uuid.uuid4().hex}"
        # shared_context.register('token_generator', TokenGenerator())

        # Example 3: Register a lazy-computed value (computed on first access)
        # shared_context.register('db_connection', lambda: create_db_connection(), lazy=True)

        # Example 4: Access values later with callable default (auto-cached)
        # import time
        # timestamp = shared_context.get('timestamp', lambda: time.time())
        # same_timestamp = shared_context.get('timestamp', lambda: time.time())  # Returns cached value

        logger.info("create_shared_context lifecycle hook completed successfully")
    except Exception as exc:
        logger.error("create_shared_context lifecycle hook failed: %s", exc, exc_info=True)
        raise
